### PR TITLE
[CALCITE-4636] Switch out of agg mode when constructing RelCollation …

### PR DESCRIPTION
**Pull request by a-rafay**
…for aggregate functions (Rafay)

### CODE REVIEW

**Comment by zinking** [comment]
LGTM. does it only generates impact on your hypothetical query? or other queries are wrong as well ?

**Comment by a-rafay** [comment]
> LGTM. does it only generates impact on your hypothetical query? or other queries are wrong as well ?

@zinking the test case i fixed in this patch was wrong. So there might be other queries giving wrong results. I found one in my test case too, but it was very similar to the test case which i fixed.

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java [hunk 1]
```diff
@@ -5758,14 +5758,21 @@ private void translateAgg(SqlCall call, @Nullable SqlNode filter,
...
-            orderList.stream()
-                .map(order ->
-                    bb.convertSortExpression(order,
-                        RelFieldCollation.Direction.ASCENDING,
-                        RelFieldCollation.NullDirection.UNSPECIFIED,
-                        this::sortToFieldCollation))
-                .collect(Collectors.toList()));
+        try {
+          // switch out of agg mode
+          bb.agg = null;
```
Comment (amaliujia) [inline core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:5763]: My question is why not keep the reference of bb.agg and then in the finally block, assign that reference back? (or is the bb.agg is always equal to this class?)

Comment (a-rafay) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:5763]: It must be equal to this class. If you scroll at the top of this function, you'll see that. We also do something similar earlier in this function as well.

Comment (a-rafay) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:5763]: See: https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java#L5585
and:
https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java#L5700

Comment (amaliujia) [reply core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:5763]: Thanks for clarifying. 

I can see now this is already an existing practice.

**Comment by a-rafay** [comment]
@amaliujia any update on this?

**Comment by amaliujia** [comment]
Fixed via https://github.com/apache/calcite/commit/4e679f03771c89ffa463051b5698ead1d935748b

**Pull request closed without merging**